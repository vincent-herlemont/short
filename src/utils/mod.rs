pub mod find;
pub mod write_all;
