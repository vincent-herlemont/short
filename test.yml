var: 1

object:
 toto: "dsa"
 tota: "dsaas"

array:
 - 1
 - 2
 - 3

array: [ 1 , 2 , 3 ]

array:
 - name: "orion"
   jambes: 2
 - name: "vincent"
   jambes: 3

