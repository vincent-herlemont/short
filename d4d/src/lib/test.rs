//! Helper for test related of d4d domain.
use crate::lib::resource;
use crate::lib::resource::Resource;
use std::fs;
use std::path::PathBuf;
use tempdir::TempDir;

pub struct Config {
    pub tmp_dir: PathBuf,
}

impl Drop for Config {
    fn drop(&mut self) {
        fs::remove_dir_all(self.tmp_dir.clone()).expect("can not clean tmp directory");
    }
}

/// Return [`InspectorConfig`], create temporary directory and copy resource on it.
///
/// The temporary directory is owned by [`InspectorConfig.path`].
///
/// # Recommendation
///
/// Need to call [`after`] at the end of test.
pub fn before(test_name: &str) -> Config {
    let test_name = format!("{}.{}", "d4d", test_name);

    // Create temporary directory.
    let path = TempDir::new(test_name.as_str())
        .expect("fail to create temporary directory")
        .into_path();

    // Copy resources to it.
    resource::to_dir(&path).expect("fail to copy resources");

    Config { tmp_dir: path }
}

/// Assert that patter value or|and expression is present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_find!(v, 2);
///assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
///assert_find!(v, el, el < &&5)
///assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}
///
/// ```
#[macro_export]
macro_rules! assert_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, true)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, true)
    };
    ($v:ident,$arms:pat,$e:expr,$b:expr) => {
        if (&$v)
            .iter()
            .find(|el| match el {
                $arms => $e,
                _ => false,
            })
            .is_some()
            != $b
        {
            panic!(format!(
                "assertion failed: {} {{{}}} {} in {{{}}}",
                if ($b) {
                    String::from("can not found")
                } else {
                    String::from("found")
                },
                String::from(stringify!($arms)),
                if (stringify!($e) == "true" || stringify!($e) == "false") {
                    String::from("")
                } else {
                    String::from(format!(
                        "with expresion {{{}}}",
                        String::from(stringify!($e))
                    ))
                },
                String::from(stringify!($v))
            ));
        };
    };
}

/// Assert that patter value or|and expression is not present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_not_find!(v, 8);
///assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
///assert_not_find!(v, el, el > &&5);
///assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
/// ```
#[macro_export]
macro_rules! assert_not_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, false)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, false)
    };
}

#[cfg(test)]
mod tests {

    #[allow(unreachable_patterns)]
    #[test]
    fn assert_find_macro_test() {
        let v = vec![1, 2, 3, 4];
        assert_find!(v, 2);
        // assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
        assert_find!(v, el, el < &&5);
        // assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}

        assert_not_find!(v, 8);
        // assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
        assert_not_find!(v, el, el > &&5);
        //assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
    }
}

pub const TEST_RESOURCE_DIRECTORY: &'static str = "./init_tpl";

/// Get all [`Resource`]
pub fn get_resource() -> Vec<Resource> {
    vec![
        Resource::new(
            "./init_tpl/other_conf.yaml",
            include_str!("./init_tpl/0_other_conf.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate.yaml",
            include_str!("./init_tpl/1_certificate.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate_altered.yaml",
            include_str!("./init_tpl/2_certificate_altered.yaml"),
        ),
        Resource::new(
            "./init_tpl/3_test/0_test.js",
            include_str!("./init_tpl/3_test/0_test.js"),
        ),
        Resource::new(
            "./init_tpl/4_tpl_certificate/certificate.yaml",
            include_str!("./init_tpl/4_tpl_certificate/certificate.yaml"),
        ),
    ]
}
