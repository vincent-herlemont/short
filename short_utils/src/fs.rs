//! File manipulation operations related of short domain.
use super::io;
use crate::error::Error;
use crate::result::{unwrap_partition, Result};
use std::cmp::Ordering;
use std::fs::File;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`PathBuf`] of files and his content.
#[derive(Debug, Eq, PartialEq)]
pub struct ContentFile {
    pub path: PathBuf,
    pub contents: String,
}

impl Ord for ContentFile {
    fn cmp(&self, other: &Self) -> Ordering {
        self.path.cmp(&other.path)
    }
}

impl PartialOrd for ContentFile {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.path.cmp(&other.path))
    }
}

impl ContentFile {
    /// Return [`ContentFile`] when the predicate [`next_line`] is satisfied.
    pub fn read_contain<F>(path: &PathBuf, next_line: F) -> Result<ContentFile>
    where
        F: Fn(&str) -> bool,
    {
        let contents = match File::open(path).map(|file| BufReader::new(file)) {
            Ok(buffer) => io::read_to_string_contain(buffer, next_line),
            Err(error) => Err(Error::from(error)),
        };

        match contents {
            Ok(contents) => Ok(ContentFile {
                path: path.to_owned(),
                contents,
            }),
            Err(error) => Err(error),
        }
    }

    /// Return each [`ContentFile`] and [`Error`] when the predicate [`next_line`] is satisfied or not.
    pub fn read_contain_multi<F>(paths: &[PathBuf], next_line: F) -> (Vec<ContentFile>, Vec<Error>)
    where
        F: Fn(&str) -> bool,
    {
        let results: (Vec<_>, Vec<_>) = paths
            .into_iter()
            .map(|path| ContentFile::read_contain(path, &next_line))
            .partition(Result::is_ok);
        unwrap_partition(results)
    }
}

#[cfg(test)]
mod tests {
    use crate::asset::default_assets;
    use crate::assets::get_all;
    use crate::error::Error;
    use crate::fs::ContentFile;
    use crate::path::retrieve;
    use crate::test::before;
    use std::io;

    #[allow(unreachable_patterns)]
    #[test]
    fn read_contain_multi_test() {
        let config = before("search_test", default_assets(get_all()));
        let paths = retrieve(&config.tmp_dir).unwrap();
        let (mut content_files, errors) =
            ContentFile::read_contain_multi(&paths, |line| line.contains("test"));

        content_files.sort();
        // ContentFiles
        assert!((&content_files).len() >= 2);
        assert_find!(
            content_files,
            ContentFile { contents, path },
            contents == &String::from("console.log(\'test.js\');")
                && path.to_string_lossy().contains("test.js")
        );

        assert!((&errors).len() > 3);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::Other);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::NotFound);
    }
}
