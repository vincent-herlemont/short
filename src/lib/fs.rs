//! File manipulation operations related of d4d domain.
use super::io;
use std::fs::File;
use std::io as sio;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`File`] and his content.
#[derive(Debug)]
struct ContentFile {
    path: PathBuf,
    contents: String,
}

fn read_to_string_finds<F>(paths: &[PathBuf], f: F) -> (Vec<ContentFile>, Vec<sio::Error>)
where
    F: Fn(&str) -> bool,
{
    let (content_files, errors): (Vec<_>, Vec<_>) = paths
        .into_iter()
        .map(|path| {
            let contents = match File::open(path).map(|file| BufReader::new(file)) {
                Ok(buffer) => io::read_to_string_finds(buffer, &f),
                Err(error) => Err(error),
            };

            match contents {
                Ok(contents) => Ok(ContentFile {
                    path: path.to_owned(),
                    contents,
                }),
                Err(error) => Err(error),
            }
        })
        .partition(Result::is_ok);

    (
        content_files.into_iter().map(Result::unwrap).collect(),
        errors.into_iter().map(Result::unwrap_err).collect(),
    )
}

#[cfg(test)]
mod tests {
    use crate::lib::fs::read_to_string_finds;
    use crate::lib::path::retrieve;
    use crate::lib::test::{after, before};

    #[test]
    fn read_to_string_finds_test() {
        let tmpath = before("search_test");
        let paths = retrieve(&tmpath).unwrap();
        let (content_files, errors) = read_to_string_finds(&paths, |line| line.contains("test"));

        // ContentFiles
        assert_eq!((&content_files).len(), 1);
        assert_eq!((&content_files)[0].contents, "console.log(\'test.js\');");
        assert!((&content_files)[0]
            .path
            .to_str()
            .unwrap()
            .contains("test.js"));

        // Errors
        assert_eq!((&errors).len(), 2);
        after(tmpath);
    }
}
