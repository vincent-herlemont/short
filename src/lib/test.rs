//! Helper for test related of d4d domain.

use std::fs;
use std::path::PathBuf;

use tempdir::TempDir;

use crate::resource;

pub struct Config {
    pub tmp_dir: PathBuf,
}

impl Drop for Config {
    fn drop(&mut self) {
        fs::remove_dir_all(self.tmp_dir.clone()).expect("can not clean tmp directory");
    }
}

/// Return [`InspectorConfig`], create temporary directory and copy resource on it.
///
/// The temporary directory is owned by [`InspectorConfig.path`].
///
/// # Recommendation
///
/// Need to call [`after`] at the end of test.
pub fn before(test_name: &str) -> Config {
    let test_name = format!("{}.{}", "d4d", test_name);

    // Create temporary directory.
    let path = TempDir::new(test_name.as_str())
        .expect("fail to create temporary directory")
        .into_path();

    // Copy resources to it.
    resource::to_dir(&path).expect("fail to copy resources");

    Config { tmp_dir: path }
}
