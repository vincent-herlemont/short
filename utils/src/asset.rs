//! Embedding and shifting of asset
use super::error::Error;
use std::collections::HashMap;
use std::error::Error as stdError;
use std::fs;
use std::path::{Path, PathBuf};

pub const ASSETS_DIRECTORY: &'static str = "assets";

#[derive(Debug)]
pub struct Asset {
    path: PathBuf,
    pub data: String,
}

impl Asset {
    pub fn new(path: &str, data: &str) -> Asset {
        Asset {
            path: PathBuf::from(path),
            data: data.to_string(),
        }
    }

    pub fn path(&self) -> &Path {
        return &self.path.as_path();
    }
}

/// Copy all [`Asset`] in target directory [`path`].
pub fn to_dir(
    path: &Path,
    assets: HashMap<&'static str, &'static str>,
) -> Result<(), Box<dyn stdError>> {
    if !path.exists() {
        return Err(Error::new_box(format!("directory {:?} not exists", path)));
    }
    for (asset_path, contents) in assets {
        let asset_path = PathBuf::from(asset_path);
        let path = path.join(asset_path.strip_prefix(ASSETS_DIRECTORY)?);
        if let Some(parent) = path.parent() {
            if !parent.exists() {
                fs::create_dir_all(parent)?;
            }
        }
        fs::write(&path, &contents)?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::assets::get_all;
    use std::fs::read_dir;
    use tempdir::TempDir;

    #[allow(unreachable_patterns)]
    #[test]
    fn copy_all_assets_to_target_directory() {
        let tempdir = TempDir::new("copy_all_assets_to_target_directory").unwrap();
        let tempdir = tempdir.path();
        to_dir(&tempdir, get_all()).unwrap();
        let files: Vec<_> = read_dir(&tempdir)
            .unwrap()
            .map(|o| o.unwrap().path())
            .collect();
        assert_find!(
            files,
            dir_entry,
            dir_entry.strip_prefix(&tempdir).unwrap() == Path::new("valid_aws_template.yaml")
        );
        assert_find!(
            files,
            dir_entry,
            dir_entry.strip_prefix(&tempdir).unwrap() == Path::new("altered_aws_template.yaml")
        );
        assert_find!(
            files,
            dir_entry,
            dir_entry.strip_prefix(&tempdir).unwrap() == Path::new("test")
        );
    }
}
