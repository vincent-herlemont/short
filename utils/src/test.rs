//! Helper for test related of d4d domain.
use crate::asset::{to_dir, Assets};
use crate::result::Result;
use std::collections::HashMap;
use std::env::current_exe;
use std::ffi::OsStr;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempdir::TempDir;
use walkdir::WalkDir;

#[derive(Debug)]
pub struct Config {
    pub tmp_dir: PathBuf,
    temp_dir: TempDir,
}

#[derive(Debug)]
pub struct ConfigCli {
    // Config
    pub tmp_dir: PathBuf,
    temp_dir: TempDir,

    pub tmp_home_dir: PathBuf,
    pub tmp_project_dir: PathBuf,
    pub tmp_private_env_dir: PathBuf,
    pub exec_path: PathBuf,
}

const HOME: &'static str = "home/.keep";
const PROJECT: &'static str = "project/.keep";
const PRIVATE_ENV: &'static str = "private_env/.keep";

impl ConfigCli {
    pub fn command(&self) -> Command {
        self.other_command(&self.exec_path)
    }

    pub fn other_command<S: AsRef<OsStr>>(&self, program: S) -> Command {
        let mut command = Command::new(program);
        command.current_dir(&self.tmp_project_dir);
        command.env("HOME", &self.tmp_home_dir);
        command
    }

    pub fn add_asset_project<P, S>(&self, path: P, content: S) -> Result<()>
    where
        P: AsRef<Path>,
        S: AsRef<str>,
    {
        let project_dir = PathBuf::from(PROJECT).parent().unwrap().to_owned();
        let path = project_dir.join(PathBuf::from(path.as_ref()));
        self.add_asset(path, content)
    }

    pub fn add_asset_home<P, S>(&self, path: P, content: S) -> Result<()>
    where
        P: AsRef<Path>,
        S: AsRef<str>,
    {
        let project_dir = PathBuf::from(HOME).parent().unwrap().to_owned();
        let path = project_dir.join(PathBuf::from(path.as_ref()));
        self.add_asset(path, content)
    }

    pub fn add_asset_private_env<P, S>(&self, path: P, content: S) -> Result<()>
    where
        P: AsRef<Path>,
        S: AsRef<str>,
    {
        let project_dir = PathBuf::from(PRIVATE_ENV).parent().unwrap().to_owned();
        let path = project_dir.join(PathBuf::from(path.as_ref()));
        self.add_asset(path, content)
    }
}

impl Config {
    pub fn cli<S: AsRef<str>>(self, carte_name: S) -> ConfigCli {
        let mut assets = HashMap::new();
        assets.insert(HOME, "");
        assets.insert(PROJECT, "");
        assets.insert(PRIVATE_ENV, "");

        to_dir(&self.tmp_dir, Assets::Static(assets)).expect("fail to copy cli assets");

        let home_dir = self.tmp_dir.join(HOME).parent().unwrap().to_path_buf();
        let project_dir = self.tmp_dir.join(PROJECT).parent().unwrap().to_path_buf();
        let private_env_dir = self
            .tmp_dir
            .join(PRIVATE_ENV)
            .parent()
            .unwrap()
            .to_path_buf();

        let current_exec = current_exe()
            .unwrap()
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .to_path_buf()
            .join(carte_name.as_ref());

        ConfigCli {
            temp_dir: self.temp_dir,
            tmp_dir: self.tmp_dir,

            tmp_home_dir: home_dir,
            tmp_project_dir: project_dir,
            tmp_private_env_dir: private_env_dir,
            exec_path: current_exec,
        }
    }
}

/// ConfigPath
pub trait ConfigPath {
    /// Return the root folder path
    fn path(&self) -> PathBuf;

    /// Return tree of all path (file and directory) relative to the root folder.
    fn tree(&self) -> Vec<PathBuf> {
        let mut tree: Vec<PathBuf> = WalkDir::new(self.path())
            .into_iter()
            .filter_map(|dir_entry| {
                if let Ok(dir_entry) = dir_entry {
                    if let Ok(path) = dir_entry.path().strip_prefix(self.path()) {
                        Some(path.to_owned())
                    } else {
                        None
                    }
                } else {
                    None
                }
            })
            .collect();
        tree.sort();
        tree
    }

    /// Allow to add assets
    fn add_asset<P, S>(&self, path: P, content: S) -> Result<()>
    where
        P: AsRef<Path>,
        S: AsRef<str>,
    {
        let mut assets = HashMap::new();
        assets.insert(PathBuf::from(path.as_ref()), String::from(content.as_ref()));
        to_dir(&self.path(), Assets::Dynamic(assets))?;
        Ok(())
    }
}

impl ConfigPath for Config {
    fn path(&self) -> PathBuf {
        self.tmp_dir.to_owned()
    }
}

impl ConfigPath for ConfigCli {
    fn path(&self) -> PathBuf {
        self.tmp_dir.to_owned()
    }
}

/// Return [`InspectorConfig`], create temporary directory and copy asset on it.
///
/// The temporary directory is owned by [`InspectorConfig.path`].
///
pub fn before(test_name: &str, assets: Assets) -> Config {
    let test_name = format!("{}.{}", "d4d", test_name);

    // Create temporary directory.
    let temp_dir = TempDir::new(test_name.as_str()).expect("fail to create temporary directory");

    // Copy assets to it.
    to_dir(temp_dir.path(), assets).expect("fail to copy assets");

    Config {
        tmp_dir: temp_dir.path().to_path_buf(),
        temp_dir,
    }
}

/// Assert that patter value or|and expression is present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// use utils::assert_find;
/// use std::panic::catch_unwind;
/// let v = vec![1,2,3,4];
///
/// assert_find!(v, 2);
/// catch_unwind(|| {
///  assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
/// });
/// assert_find!(v, el, el < &&5);
/// catch_unwind(|| {
///  assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}
/// });
///
/// ```
#[macro_export]
macro_rules! assert_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, true)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, true)
    };
    ($v:ident,$arms:pat,$e:expr,$b:expr) => {
        if (&$v)
            .iter()
            .find(|el| match el {
                $arms => $e,
                _ => false,
            })
            .is_some()
            != $b
        {
            panic!(format!(
                "assertion failed: {} {{{}}} {} in {{{}}}",
                if ($b) {
                    String::from("can not found")
                } else {
                    String::from("found")
                },
                String::from(stringify!($arms)),
                if (stringify!($e) == "true" || stringify!($e) == "false") {
                    String::from("")
                } else {
                    String::from(format!(
                        "with expresion {{{}}}",
                        String::from(stringify!($e))
                    ))
                },
                String::from(stringify!($v))
            ));
        };
    };
}

/// Assert that patter value or|and expression is not present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// use std::panic::catch_unwind;
/// use crate::utils::assert_not_find;
/// use crate::utils::assert_find; // TODO : found a way to not have to import assert_find macro.
/// let v = vec![1,2,3,4];
///
/// assert_not_find!(v, 8);
/// catch_unwind(|| {
///  assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
/// });
/// assert_not_find!(v, el, el > &&5);
/// catch_unwind(|| {
///  assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
/// });
/// ```
#[macro_export]
macro_rules! assert_not_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, false)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, false)
    };
}

#[cfg(test)]
mod tests {

    #[allow(unreachable_patterns)]
    #[test]
    fn assert_find_macro_test() {
        let v = vec![1, 2, 3, 4];
        assert_find!(v, 2);
        // assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
        assert_find!(v, el, el < &&5);
        // assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}

        assert_not_find!(v, 8);
        // assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
        assert_not_find!(v, el, el > &&5);
        //assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
    }
}
