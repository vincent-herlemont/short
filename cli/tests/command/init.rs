use insta::assert_debug_snapshot;
use std::fs::read_to_string;
use utils::asset::Assets;
use utils::test::{before, ConfigPath};

const CRATE_NAME: &'static str = env!("CARGO_PKG_NAME");

#[test]
fn init() {
    let config = before("init", Assets::None).cli(CRATE_NAME);
    let output = config.command().arg("init").output().unwrap();

    assert_eq!("\n", String::from_utf8(output.stdout).unwrap());

    let content_dir = config.tree();
    assert_debug_snapshot!(content_dir);

    let local_project_file = &config.tmp_project_dir.join("short.yaml");
    let content = read_to_string(local_project_file).unwrap();
    assert_eq!(
        r#"---
projects: []"#,
        content.as_str()
    );

    // project file must not to be modify
    const PROJECT_FILE_CONTENT: &'static str = r#"---
projects:
  - name: p1
    public_env_directory: "."
    provider:
        name: aws
"#;
    config
        .add_asset_project("./short.yaml", PROJECT_FILE_CONTENT)
        .unwrap();
    let _output = config.command().arg("init").output().unwrap();
    let local_project_file = &config.tmp_project_dir.join("short.yaml");
    let content = read_to_string(local_project_file).unwrap();
    assert_eq!(PROJECT_FILE_CONTENT, content.as_str());
}
