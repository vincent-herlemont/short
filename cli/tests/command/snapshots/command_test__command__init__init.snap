---
source: cli/tests/command/init.rs
expression: content_dir
---
[
    "",
    "home",
    "home/.keep",
    "private_env",
    "private_env/.keep",
    "project",
    "project/.keep",
    "project/short.yaml",
]
